### DISCUSSION TITLE
NativeControlHost Non-Windows Platforms

### DISCUSSION THREAD

**Discussion by djrecipe** [AvaloniaUI/Avalonia#11069] [Q&A] [answered]
Hello all,

I'm currently using NativeControlHost to host natively rendered OpenGL content in an Avalonia user control.

I'd like to explore supporting platforms other than Windows, and I was hoping someone had the expertise to answer a few quick questions:
- Is it possible to render native (C++) content in an Avalonia control on any non-Windows platform. (I realize both window handles and OpenGL work a bit differently on other platforms)
- Is there a NativeControlHost equivalent for other platforms and/or does NativeControlHost support other platforms
- Does Avalonia have any support for retrieving the handle of a user control (or its parent window) on a non-Window platform
- Is it possible to render native content in an Avalonia control on a Non-Windows platform (I'd prefer to write native OpenGL code for each platform)

Thanks a ton for anyone who may have time to help answer my inquiries. Thus far I've found Avalonia to be largely superior option to MAUI for my needs.

Thank you,
Justin Pierce

**Comment by jp2masa** [comment, marked as answer]
There is a demo in ControlCatalog:

https://github.com/AvaloniaUI/Avalonia/blob/532e89c837d61d37d7fb2273f852d51bdc0f1999/samples/ControlCatalog/Pages/NativeEmbedPage.xaml

https://github.com/AvaloniaUI/Avalonia/blob/532e89c837d61d37d7fb2273f852d51bdc0f1999/samples/ControlCatalog/Pages/NativeEmbedPage.xaml.cs

With implementations for:

- Win: https://github.com/AvaloniaUI/Avalonia/blob/532e89c837d61d37d7fb2273f852d51bdc0f1999/samples/ControlCatalog.NetCore/NativeControls/Win/EmbedSample.Win.cs
- Mac: https://github.com/AvaloniaUI/Avalonia/blob/532e89c837d61d37d7fb2273f852d51bdc0f1999/samples/ControlCatalog.NetCore/NativeControls/Mac/EmbedSample.Mac.cs
- Gtk: https://github.com/AvaloniaUI/Avalonia/blob/532e89c837d61d37d7fb2273f852d51bdc0f1999/samples/ControlCatalog.NetCore/NativeControls/Gtk/EmbedSample.Gtk.cs
- Android: https://github.com/AvaloniaUI/Avalonia/blob/532e89c837d61d37d7fb2273f852d51bdc0f1999/samples/ControlCatalog.Android/EmbedSample.Android.cs
- iOS: https://github.com/AvaloniaUI/Avalonia/blob/532e89c837d61d37d7fb2273f852d51bdc0f1999/samples/ControlCatalog.iOS/EmbedSample.iOS.cs
- Web: https://github.com/AvaloniaUI/Avalonia/blob/532e89c837d61d37d7fb2273f852d51bdc0f1999/samples/ControlCatalog.Browser/EmbedSample.Browser.cs

**Reply by djrecipe** [reply]
You're incredible, can't wait to check this out and parse the examples! 🥳
